File: TIC_TAC_TOE.java
import java.util.*;
class TIC_TAC_TOE
{
    char p[]=new char[2],a[]=new char[9];
    TIC_TAC_TOE()
    {
        for(int i=0;i<9;++i)
            a[i]='-';
    }

    private void display()
    {
        System.out.println("\fWelcome to play TIC-TAC-TOE world\nfor reference to players");
        System.out.println(" ____________________");
        System.out.println("| a[0] | a[1] | a[2] |");
        System.out.println("|______|______|______|");
        System.out.println("| a[3] | a[4] | a[5] |");
        System.out.println("|______|______|______|");
        System.out.println("| a[6] | a[7] | a[8] |");
        System.out.println("|______|______|______|");
    }

    void input()
    {
        System.out.println("Enter X or O for Player1");
        p[0]=Character.toUpperCase(new Scanner(System.in).next().charAt(0));
        assign();
    }

    private void assign()
    {
        switch(p[0])
        {
            case 'X':p[1]='O';
            break;
            case 'O':p[1]='X';
            break;
            default:System.out.println("Invalid Choice!!!");
            input();
            return;
        }
        System.out.println("Player2 choice is "+p[1]);
    }

    private boolean check(int a,char ch)
    {
        int b=a%3;
        int c=0;
        for(int i=1;i<=3;++i,b+=3)
            if(this.a[b]==ch)
                ++c;
        if(c==3)
            return true;
        b=3*(a/3);
        c=0;
        for(int i=1;i<=3;++i,++b)
            if(this.a[b]==ch)
                ++c;
        if(c==3)
            return true;
        c=0;
        if(a%2==0)
        {
            if(a!=4)
            {
                int in;
                if(a%4==0)
                    in=4;
                else
                    in=2;
                b=a%4;
                for(int i=1;i<=3;++i,b+=in)
                    if(this.a[b]==ch)
                        ++c;
                if(c==3)
                    return true;
            }
            else
            {
                b=0;
                for(int i=1;i<=3;++i,b+=4)
                    if(this.a[b]==ch)
                        ++c;
                if(c==3)
                    return true;
                c=0;
                b=2;
                for(int i=1;i<=3;++i,b+=2)
                    if(this.a[b]==ch)
                        ++c;
                if(c==3)
                    return true;
            }
        }
        return false;
    }

    private void printv()
    {
        int x=0;
        for(int i=0;i<3;++i)
        {
            for(int j=0;j<3;++j)
                System.out.print("\t"+a[x++]);
            System.out.println();
        }
    }

    private void startGame()
    {
        Scanner sc=new Scanner(System.in);
        boolean flag=true;
        for(int i=1;i<=9;++i)
        {
            int ind=2-i%2;
            System.out.println("Player"+ind+" Enter the position");
            int pos=sc.nextInt();
            a[pos]=p[ind-1];
            printv();
            if(check(pos,p[ind-1]))
            {
                System.out.println("Player"+ind+" wins");
                flag=false;
                break;
            }
        }
        if(flag)
            System.out.println("None wins");
        System.out.println("Thank you for playing this game");
    }

    public static void main(String args[])
    {
        TIC_TAC_TOE o1=new TIC_TAC_TOE();
        o1.display();
        o1.input();
        o1.startGame();
    }
}

File: TIC-TAC-TOE/TIC-TAC-TOE.py
players=['','X','O']
Board=[' ']*10
score=['',0,0]


from IPython.display import clear_output

def display_board(board):
    
    print(f'   |   |')
    print(f' {board[7]} | {board[8]} | {board[9]}')
    print(f'---|---|---')
    print(f' {board[4]} | {board[5]} | {board[6]}')
    print(f'---|---|---')
    print(f' {board[1]} | {board[2]} | {board[3]}')
    print(f'   |   |')


def player_input():
    play1=' '
    while play1 not in  ('X','O'):
        play1=input('Player 1 : Choose either \'X\' or \'O\' : ').upper()
    if play1=='X':
        play2='O'
    else:
        play2='X'
    return (play1,play2)



def place_marker(board, marker, position):
    board[position]=marker



def win_check(board, mark):
    return ((board[1]==board[2]==board[3]==mark) or 
    (board[4] ==  board[5] ==  board[6] == mark) or 
    (board[7] ==  board[8] ==  board[9] == mark) or
    (board[7] ==  board[4] ==  board[1] == mark) or 
    (board[8] ==  board[5] ==  board[2] == mark) or
    (board[9] ==  board[6] ==  board[3] == mark) or
    (board[7] ==  board[5] ==  board[3] == mark) or
    (board[9] ==  board[5] ==  board[1] == mark))



import random

def choose_first():
    return random.choice((-1,1))



def space_check(board, position):
    return board[position]==' '



def full_board_check(board):
    return not ' ' in board[1:]


def player_choice(board,player):
    pos = 0
    while pos not in range(1,11) or not space_check(board,pos):
        try:
            pos = int(input(f'Player{player} Enter the position : '))
        except:
            print('Sorry try again')
    return pos


def replay():
    return input('Do you want to play again (Yes/No) : ').lower()=='yes'



def gridfor():
    print(f'   |   |')
    print(f' 7 | 8 | 9')
    print(f'---|---|---')
    print(f' 4 | 5 | 6')
    print(f'---|---|---')
    print(f' 1 | 2 | 3')
    print(f'   |   |')


def scorepr():
    print (f'Player 1 : {score[1]}')
    print (f'Player 2 : {score[2]}')
    
    

print('Welcome to Tic Tac Toe!')
print ('\ngrid format : ')
gridfor()
x=1
while True:
    print ('\nSCORE BOARD :')
    scorepr()
    ind = choose_first()
    player = players[ind]
    print(f'This is round {x}, Player {player} will go first!')
    ongoing = True
    input('Hit Enter to continue')
    while ongoing:
        display_board(Board)
        print('\n')
        gridfor()
        position = player_choice(Board,player)
        place_marker(Board,player,position)
        
        if win_check(Board, player):
            display_board(Board)
            print('Congratulations! Player '+player+' wins!')
            score[ind]+=1 
            ongoing = False
        else:
            if full_board_check(Board):
                display_board(Board)
                print('The game is a draw!')
                break
            else:
                ind *= -1
                player = players[ind]
                clear_output()
                
    Board = [' '] * 10
    x+=1
    
    if not replay():
        break
print('\nFINAL SCORE :')
scorepr()
score=['',0,0]

